File: mict_framework_python/mict_framework.py
# mict-framework.py

import time
import threading
from typing import Callable, Dict, List, Optional, Any, Union

class MICTError(Exception):
    """Base class for MICT-related errors."""
    pass

class ConfigurationError(MICTError):
    """Raised for errors in the MICT configuration."""
    pass

class MICT:
    """
    A class implementing the Mobius Inspired Cyclical Transformation (MICT) framework.

    Args:
        config (dict): A dictionary containing configuration options for the MICT cycle.

    Attributes:
        stages (list): A list of stage names (strings) defining the MICT cycle.
        current_state (dict): The current state of the system.
        previous_state (dict): The previous state of the system.
        current_stage_index (int): The index of the current stage in the `stages` list.
        updateUI (callable): A function to call for updating the user interface.
        stage_functions (dict): A dictionary mapping stage names to functions.
        interval_id (threading.Timer or None): The ID of the background thread (if running).
        interval (int): The interval (in milliseconds) between cycle iterations.
        error_handler (callable or None): A function to call for handling errors.
        config (dict):  Stores the original configuration
    """
    def __init__(self, config: Dict):
        if not isinstance(config, dict):
            raise TypeError("MICT constructor requires a configuration object.")
        if not isinstance(config.get('stages'), list) or not config['stages']:
            raise TypeError("MICT configuration must include a 'stages' list with at least one stage.")
        if not callable(config.get('updateUI')):
            raise TypeError("MICT configuration must include an 'updateUI' function.")

        self.stages: List[str] = config['stages']
        self.current_state: Dict = config.get('initialState', {})  # Default to empty dict if not provided
        self.previous_state: Optional[Dict] = None  # Initialize previous_state
        self.current_stage_index: int = 0
        self.updateUI: Callable[[Dict, str], None] = config['updateUI']
        self.stage_functions: Dict[str, Callable] = config.get('stageFunctions', {})  # Default to empty dict
        self.interval_id: Optional[threading.Timer] = None
        self.interval: int = config.get('interval', 0)  # Default to 0 (no interval)
        self.error_handler: Optional[Callable[[Exception, str, Dict], None]] = config.get('errorHandler', None)
        self._stop_event: threading.Event = threading.Event() # Event for stopping

        # Validate stageFunctions
        for stage_name, func in self.stage_functions.items():
            if stage_name not in self.stages:
                raise TypeError(f"Invalid stage name '{stage_name}' in stageFunctions. Must be one of: {', '.join(self.stages)}")
            if not callable(func):
                raise TypeError(f"stageFunctions['{stage_name}'] must be a function.")

        self.config = config #Store for later use.

    def get_current_stage(self) -> str:
        """Returns the name of the current stage."""
        return self.stages[self.current_stage_index]

    def get_current_stage_index(self) -> int:
        """Returns the index of the current stage."""
        return self.current_stage_index
    
    def get_previous_state(self) -> Optional[Dict]:
        """Returns the previous state of the system."""
        return self.previous_state

    def next_stage(self) -> None:
        """Executes the next stage in the MICT cycle."""
        current_stage = self.get_current_stage()

        # Call the stage function (if it exists)
        if current_stage in self.stage_functions:
            try:
                # Call stage function, potentially updating state
                new_state = self.stage_functions[current_stage](self.current_state)
                self.previous_state = self.current_state.copy() # Use copy to avoid modifying.
                if new_state is not None:  # Allow for no change.
                    self.current_state = new_state
            except Exception as error:
                if self.error_handler:
                    self.error_handler(error, current_stage, self.current_state)
                else:
                    print(f"Error in stage '{current_stage}': {error}")
                return  # Don't proceed to the next stage if there's an error

        # Move to the next stage
        self.current_stage_index = (self.current_stage_index + 1) % len(self.stages)
        self.updateUI(self.current_state, self.get_current_stage())

    def start_cycle(self, interval: Optional[int] = None) -> Optional[threading.Thread]:
        """
        Starts the MICT cycle.

        Args:
            interval (int, optional): The interval (in milliseconds) between cycle iterations.
                If not provided, uses the interval specified in the config.  If 0 or None,
                the cycle will not run automatically; `next_stage` must be called manually.

        Returns:
            threading.Thread or None: The thread object if the cycle is started with an interval,
                otherwise None.
        """
        if self.interval_id:
            print("MICT cycle is already running.  Stopping the previous cycle.")
            self.stop_cycle()

        self._stop_event.clear()  # Clear the stop event
        _interval = interval if interval is not None else self.interval
        if not _interval:
            print("MICT cycle started without an interval.  Call next_stage() manually.")
            return None

        def run_cycle():
            while not self._stop_event.is_set():  # Use the event to check for stopping
                self.next_stage()
                time.sleep(_interval / 1000)

        self.interval_id = threading.Thread(target=run_cycle)
        self.interval_id.daemon = True
        self.interval_id.start()
        return self.interval_id

    def stop_cycle(self) -> None:
        """Stops the MICT cycle."""
        self._stop_event.set()  # Signal the thread to stop
        self.interval_id = None


    def set_state(self, new_state: Dict) -> None:
        """
        Sets the current state of the system.

        Args:
            new_state (dict): The new state.
        """
        self.previous_state = self.current_state.copy() # Save previous state
        self.current_state = new_state
        self.updateUI(self.current_state, self.get_current_stage())

    def reset(self) -> None:
        """Resets the MICT cycle to its initial state."""
        self.current_stage_index = 0
        self.previous_state = None #Reset previous state
        self.current_state = self.config['initialState'].copy()  # Use initial state from config
        self.stop_cycle()
        self.updateUI(self.current_state, self.get_current_stage())

File: src/mict-framework.js
// mict-framework.js

/**
 * Represents a single MICT (Mobius Inspired Cyclical Transformation) cycle.
 */
class MICT {
    /**
     * Creates a new MICT instance.
     * @param {object} config - Configuration object for the MICT cycle.
     * @param {string[]} config.stages - An array of stage names (e.g., ["Mapping", "Iteration", "Checking", "Transformation"]).
     * @param {object} config.initialState - The initial state of the system.  Can be any JavaScript object.
     * @param {function} config.updateUI - A function that will be called to update the user interface (or other external systems) after each stage transition.
     * @param {object} [config.stageFunctions={}] - An optional object mapping stage names to functions.  These functions will be called *before* the stage transition.
     * @param {number} [config.interval=0] - The interval in milliseconds for automatic cycle execution. 0 (or null/undefined) disables automatic cycling.
     * @param {function} [config.errorHandler] - An optional error handler function.
     * @throws {Error} If the configuration is invalid.
     */
    constructor(config) {
        if (!config || typeof config !== 'object') {
            throw new Error("MICT constructor requires a configuration object.");
        }
        if (!Array.isArray(config.stages) || config.stages.length === 0) {
            throw new Error("MICT configuration must include a 'stages' array with at least one stage.");
        }
        if (typeof config.updateUI !== 'function') {
            throw new Error("MICT configuration must include an 'updateUI' function.");
        }

        this.stages = config.stages;
        this.currentState = config.initialState;
        this.currentStageIndex = 0;
        this.updateUI = config.updateUI;
        this.stageFunctions = config.stageFunctions || {}; // Use an empty object if not provided
        this.intervalId = null;
        this.interval = config.interval || 0; // Store the interval
        this.errorHandler = config.errorHandler || null;
		  this.previousState = null; //Add for tracking.

        // Validate stageFunctions
        for (const stageName in this.stageFunctions) {
            if (!this.stages.includes(stageName)) {
                throw new Error(`Invalid stage name '${stageName}' in stageFunctions.  Must be one of: ${this.stages.join(', ')}`);
            }
            if (typeof this.stageFunctions[stageName] !== 'function') {
                throw new Error(`stageFunctions['${stageName}'] must be a function.`);
            }
        }
    }

    /**
     * Gets the name of the current stage.
     * @returns {string} The name of the current stage.
     */
    getCurrentStage() {
        return this.stages[this.currentStageIndex];
    }
    /**
     * Gets the current numerical index
     * @returns {integer} The index of the current stage.
     */
    getCurrentStageIndex(){
        return this.currentStageIndex;
    }

    /**
     * Transitions to the next stage in the cycle.  Calls the stage function (if any) before transitioning.
     */
     nextStage() {
        const currentStage = this.getCurrentStage();

        // Call the stage function (if it exists)
        if (this.stageFunctions[currentStage]) {
            try {
                // Pass the current state to the stage function.  The stage function
                // can modify the state directly, or return a *new* state.
                const newState = this.stageFunctions[currentStage](this.currentState);
					 this.previousState = this.currentState;// Save 
                if (newState !== undefined) {
                    this.currentState = newState; // Update state if a new state is returned
                }
            } catch (error) {
                if (this.errorHandler) {
                    this.errorHandler(error, currentStage, this.currentState);
                } else {
                    console.error(`Error in stage '${currentStage}':`, error);
                    // You might want to stop the cycle here in a production environment
                    // this.stopCycle();
                }
                return; // Don't proceed to the next stage if there's an error
            }
        }

        // Move to the next stage
        this.currentStageIndex = (this.currentStageIndex + 1) % this.stages.length;
        this.updateUI(this.currentState, this.getCurrentStage()); // Pass current state and stage
    }

    /**
     * Starts the automatic execution of the MICT cycle.
     * @param {number} [interval] - The interval in milliseconds.  If not provided, uses the interval from the constructor.
     */
    startCycle(interval) {
        if (this.intervalId) {
            console.warn("MICT cycle is already running.  Stopping the previous cycle.");
            this.stopCycle();
        }

        const intervalToUse = interval || this.interval; // Use provided interval, or constructor interval
        if (!intervalToUse) {
            console.warn("MICT cycle started without an interval.  Call nextStage() manually.");
            return; // Don't start if interval is 0, null, or undefined.
        }

        this.intervalId = setInterval(() => {
            this.nextStage();
        }, intervalToUse);
    }

    /**
     * Stops the automatic execution of the MICT cycle.
     */
    stopCycle() {
        if (this.intervalId) {
            clearInterval(this.intervalId);
            this.intervalId = null;
        }
    }

    /**
     * Sets the current state of the MICT cycle.
     * @param {object} newState The new state.
     */
    setState(newState) {
        this.currentState = newState;
        this.updateUI(this.currentState, this.getCurrentStage()); // Update UI with new state
    }
    /**
    * Gets the previousState
    */
    getPreviousState(){
        return this.previousState;
    }
     /**
     * Resets the MICT cycle to its initial state and stage.
     */
    reset() {
        this.currentStageIndex = 0;
        this.currentState = this.initialState; //Should be set.
		  this.previousState = null; //Reset this as well
        this.stopCycle(); // Stop any running interval
        this.updateUI(this.currentState, this.getCurrentStage()); // Update the UI
    }
}

export { MICT };
